src/ostree/ot-builtin-prune.c

```diff
@@ -200,26 +171,112 @@ ostree_builtin_prune (int argc, char **argv, GCancellable *cancellable, GError *
         else if (!delete_commit (repo, opt_delete_commit, cancellable, error))
           goto out;
     }
-  if (opt_keep_younger_than)
-    {
-      if (opt_no_prune)
-        {
-          ot_util_usage_error (context, "Cannot specify both --keep-younger-than and --no-prune", error);
-          goto out;
-        }
-      if (!prune_commits_keep_younger_than_date (repo, opt_keep_younger_than, cancellable, error))
-        goto out;
-    }
 
   if (opt_refs_only)
     pruneflags |= OSTREE_REPO_PRUNE_FLAGS_REFS_ONLY;
   if (opt_no_prune)
     pruneflags |= OSTREE_REPO_PRUNE_FLAGS_NO_PRUNE;
 
-  if (!ostree_repo_prune (repo, pruneflags, opt_depth,
-                          &n_objects_total, &n_objects_pruned, &objsize_total,
-                          cancellable, error))
-    goto out;
+  /* If no newer more complex options are specified, drop down to the original
+   * prune API - both to avoid code duplication, and to keep it run from the
+   * test suite.
+   */
+  if (!(opt_retain_branch_depth || opt_keep_younger_than))
+    {
+      if (!ostree_repo_prune (repo, pruneflags, opt_depth,
+                              &n_objects_total, &n_objects_pruned, &objsize_total,
+                              cancellable, error))
+        goto out;
+    }
+  else
+    {
+      g_autoptr(GHashTable) all_refs = NULL;
+      g_autoptr(GHashTable) reachable = ostree_repo_traverse_new_reachable ();
+      g_autoptr(GHashTable) retain_branch_depth = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+      struct timespec keep_younger_than_ts;
+      GHashTableIter hash_iter;
+      gpointer key, value;
+
+      /* Otherwise, the default is --refs-only; we set this just as a note */
+      opt_refs_only = TRUE;
+
+      if (opt_keep_younger_than)
+        {
+          if (!parse_datetime (&keep_younger_than_ts, opt_keep_younger_than, NULL))
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "Could not parse '%s'", opt_keep_younger_than);
+              goto out;
+            }
+        }
+
+      for (char **iter = opt_retain_branch_depth; iter && *iter; iter++)
+        {
+          /* bd should look like BRANCH=DEPTH where DEPTH is an int */
+          const char *bd = *iter;
+          const char *eq = strchr (bd, '=');
+
+          if (!eq)
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "Invalid value %s, must specify BRANCH=DEPTH",
+                           bd);
+              goto out;
+            }
+          g_hash_table_insert (retain_branch_depth, g_strndup (bd, eq - bd),
+                               GINT_TO_POINTER ((int)g_ascii_strtoll (eq + 1, NULL, 10)));
```

Comment:
Since this is directly user-facing, we should probably validate that `g_ascii_strtoll` converted successfully. Otherwise, a typo can end up pruning the whole branch.